File: frontend/plugins/northrou-native/ios/Sources/NorthrouNative/NorthrouNativePlugin.swift
// Northrou's native iOS chrome.
//
// Tauri is built for "web UI + native capabilities", not native navigation
// wrapping web screens, so the split is deliberate: every screen is web, and
// the things a WebView genuinely cannot do well are native. On iOS that is the
// tab bar (a web imitation never feels right: no blur, no haptics, wrong
// safe-area behaviour, wrong swipe) and, shortly, the player.
//
// The design here is not invented: it mirrors frontend/swift/ContentView.swift
// item for item -- same titles, same SF Symbols, same order, same search role.
// That file stays the design reference. A SwiftUI TabView renders a UITabBar
// underneath anyway, so building the UITabBar directly is the same chrome
// without hosting a SwiftUI tree that owns no content.
//
// This lives in the plugin's own package on purpose. src-tauri/gen/apple is
// regenerated, and anything edited there is lost; a plugin subview survives.

import UIKit
import WebKit
import Tauri

/// A tab, mirroring the `Tabs` enum in ContentView.swift.
struct TabItem: Decodable {
    let key: String      // "home" | "films" | "shows" | "search"
    let title: String
    let systemImage: String
}

struct SetTabsArgs: Decodable {
    let tabs: [TabItem]
    let selected: String?
}

struct SetTabArgs: Decodable {
    let key: String
}

struct ShowChromeArgs: Decodable {
    let visible: Bool
}

class NorthrouNativePlugin: Plugin, UITabBarDelegate {
    private var tabBar: UITabBar?
    private var keys: [String] = []
    private weak var webView: WKWebView?

    override func load(webview: WKWebView) {
        self.webView = webview
    }

    /// Build (or rebuild) the tab bar and pin it to the bottom.
    @objc public func set_tabs(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(SetTabsArgs.self)

        DispatchQueue.main.async {
            guard let host = self.manager.viewController?.view else {
                invoke.reject("no host view controller")
                return
            }

            self.tabBar?.removeFromSuperview()
            self.keys = args.tabs.map(\.key)

            let bar = UITabBar()
            bar.delegate = self
            bar.translatesAutoresizingMaskIntoConstraints = false
            bar.items = args.tabs.enumerated().map { i, t in
                let item = UITabBarItem(
                    title: t.title,
                    image: UIImage(systemName: t.systemImage),
                    tag: i
                )
                return item
            }

            if let selected = args.selected, let i = self.keys.firstIndex(of: selected) {
                bar.selectedItem = bar.items?[i]
            } else {
                bar.selectedItem = bar.items?.first
            }

            host.addSubview(bar)
            NSLayoutConstraint.activate([
                bar.leadingAnchor.constraint(equalTo: host.leadingAnchor),
                bar.trailingAnchor.constraint(equalTo: host.trailingAnchor),
                // Pin to the view, not the safe area: a UITabBar draws its own
                // background down through the home indicator and lays its items
                // out above it. Pinning to the safe area leaves a strip of page
                // showing underneath.
                bar.bottomAnchor.constraint(equalTo: host.bottomAnchor),
            ])
            self.tabBar = bar

            // The web page must not draw its last row underneath the bar.
            self.applyInset(height: bar.intrinsicContentSize.height)
            invoke.resolve()
        }
    }

    /// Select a tab from the web side, so the bar follows in-page navigation
    /// (tapping a card, going back) instead of drifting out of sync.
    @objc public func set_tab(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(SetTabArgs.self)
        DispatchQueue.main.async {
            guard let bar = self.tabBar,
                  let i = self.keys.firstIndex(of: args.key),
                  let items = bar.items, i < items.count else {
                invoke.resolve()
                return
            }
            bar.selectedItem = items[i]
            invoke.resolve()
        }
    }

    /// Hide the chrome for immersive content (a detail modal, playback).
    @objc public func show_chrome(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(ShowChromeArgs.self)
        DispatchQueue.main.async {
            guard let bar = self.tabBar else {
                invoke.resolve()
                return
            }
            UIView.animate(withDuration: 0.2) {
                bar.alpha = args.visible ? 1 : 0
            } completion: { _ in
                bar.isHidden = !args.visible
            }
            self.applyInset(height: args.visible ? bar.intrinsicContentSize.height : 0)
            invoke.resolve()
        }
    }

    func tabBar(_ tabBar: UITabBar, didSelect item: UITabBarItem) {
        guard item.tag < keys.count else { return }
        // The web side listens for this and swaps route. The native bar owns
        // the selection; the web owns the content.
        trigger("tabChanged", data: ["key": keys[item.tag]])
    }

    /// Tell the page how much room the bar takes, as a CSS variable the layout
    /// can add to its own padding.
    private func applyInset(height: CGFloat) {
        let js = "document.documentElement.style.setProperty('--native-tabbar-height', '\(Int(height))px')"
        webView?.evaluateJavaScript(js)
    }
}

@_cdecl("init_plugin_northrou_native")
func initPlugin() -> Plugin {
    return NorthrouNativePlugin()
}

File: frontend/plugins/northrou-native/android/src/main/java/sh/northrou/chrome/NorthrouNativePlugin.kt
// Northrou's native Android chrome.
//
// The mirror of the iOS plugin, and deliberately not shared with it: per
// docs/frontend.md, native chrome is written twice. That is the price of a real
// UITabBar on iOS and a real BottomNavigationView here, instead of a web
// imitation that is subtly wrong on both.
//
// Kept in the plugin's own package, not in the generated gen/android project,
// which is regenerated and would lose it.

package sh.northrou.chrome

import android.app.Activity
import android.view.Gravity
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.FrameLayout
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import com.google.android.material.bottomnavigation.BottomNavigationView

@InvokeArg
class TabItem {
    lateinit var key: String
    lateinit var title: String

    // The iOS side names SF Symbols here. Android resolves its own drawable by
    // tab key rather than trying to map Apple's symbol names onto material
    // icons, which never lines up.
    var systemImage: String = ""
}

@InvokeArg
class SetTabsArgs {
    var tabs: List<TabItem> = emptyList()
    var selected: String? = null
}

@InvokeArg
class SetTabArgs {
    lateinit var key: String
}

@InvokeArg
class ShowChromeArgs {
    var visible: Boolean = true
}

@TauriPlugin
class NorthrouNativePlugin(private val activity: Activity) : Plugin(activity) {
    private var nav: BottomNavigationView? = null
    private var keys: List<String> = emptyList()
    private var webView: WebView? = null

    override fun load(webView: WebView) {
        this.webView = webView
    }

    private fun iconFor(key: String): Int = when (key) {
        "home" -> android.R.drawable.ic_menu_view
        "films" -> android.R.drawable.ic_menu_gallery
        "shows" -> android.R.drawable.ic_menu_sort_by_size
        "search" -> android.R.drawable.ic_menu_search
        else -> android.R.drawable.ic_menu_help
    }

    @Command
    fun set_tabs(invoke: Invoke) {
        val args = invoke.parseArgs(SetTabsArgs::class.java)

        activity.runOnUiThread {
            val root = activity.window.decorView.findViewById<ViewGroup>(android.R.id.content)
            nav?.let { root.removeView(it) }
            keys = args.tabs.map { it.key }

            val bar = BottomNavigationView(activity)
            args.tabs.forEachIndexed { i, t ->
                bar.menu.add(0, i, i, t.title).setIcon(iconFor(t.key))
            }
            bar.setOnItemSelectedListener { item ->
                // The native bar owns the selection; the web owns the content.
                trigger("tabChanged", JSObject().put("key", keys[item.itemId]))
                true
            }
            args.selected?.let { sel ->
                keys.indexOf(sel).takeIf { it >= 0 }?.let { bar.selectedItemId = it }
            }

            val lp = FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            ).apply { gravity = Gravity.BOTTOM }

            root.addView(bar, lp)
            nav = bar

            bar.post { applyInset(bar.height) }
            invoke.resolve()
        }
    }

    @Command
    fun set_tab(invoke: Invoke) {
        val args = invoke.parseArgs(SetTabArgs::class.java)
        activity.runOnUiThread {
            keys.indexOf(args.key).takeIf { it >= 0 }?.let { nav?.selectedItemId = it }
            invoke.resolve()
        }
    }

    @Command
    fun show_chrome(invoke: Invoke) {
        val args = invoke.parseArgs(ShowChromeArgs::class.java)
        activity.runOnUiThread {
            val bar = nav
            if (bar == null) {
                invoke.resolve()
                return@runOnUiThread
            }
            bar.animate().alpha(if (args.visible) 1f else 0f).setDuration(200).withEndAction {
                bar.visibility = if (args.visible) android.view.View.VISIBLE else android.view.View.GONE
            }
            applyInset(if (args.visible) bar.height else 0)
            invoke.resolve()
        }
    }

    /// Tell the page how much room the bar takes, as a CSS variable its layout
    /// can add to its own padding.
    private fun applyInset(heightPx: Int) {
        val density = activity.resources.displayMetrics.density
        val cssPx = (heightPx / density).toInt()
        webView?.evaluateJavascript(
            "document.documentElement.style.setProperty('--native-tabbar-height', '${cssPx}px')",
            null,
        )
    }
}
